touchhttp/bundle.go

```diff
@@ -0,0 +1,392 @@
+// specific to that server.
+func (b bundle) curry(l prometheus.Labels) instrumenter {
+	i := instrumenter{
+		counter:     b.counter.MustCurryWith(l),
+		inFlight:    b.inFlight.With(l),
+		requestSize: b.requestSize.MustCurryWith(l),
+		duration:    b.duration.MustCurryWith(l),
+		now:         b.now,
+	}
+
+	if b.errorCount != nil {
+		i.errorCount = b.errorCount.MustCurryWith(l)
+	}
+
+	return i
+}
+
+// instrumenter is the common logic that decorates HTTP transactions for
+// both clients and servers.
+type instrumenter struct {
+	counter     *prometheus.CounterVec
+	inFlight    prometheus.Gauge
+	requestSize prometheus.ObserverVec
+	duration    prometheus.ObserverVec
+
+	// only used in clients
+	errorCount *prometheus.CounterVec
+
+	now func() time.Time
+}
+
+// begin records the start of an HTTP transaction
+func (i instrumenter) begin(r *http.Request) transaction {
+	i.inFlight.Inc()
+	return transaction{
+		start:       i.now(),
+		method:      cleanMethod(r.Method),
+		requestSize: r.ContentLength,
+	}
+}
+
+func (i instrumenter) endHandle(sc observe.StatusCoder, t transaction) {
+	t.code = sc.StatusCode()
+	if t.code == 0 {
+		// this can happen if a decorated handler never wrote a status code
+		t.code = http.StatusOK
+	}
+
+	i.end(t)
+}
+
+func (i instrumenter) endDo(response *http.Response, err error, t transaction) {
+	if response != nil {
+		t.code = response.StatusCode
+	} else {
+		t.code = -1
+	}
+
+	t.err = err
+	i.end(t)
+}
+
+// end records the end of an HTTP transaction
+func (i instrumenter) end(t transaction) {
+	i.inFlight.Dec()
+
+	l := prometheus.Labels{
+		MethodLabel: t.method,
+		CodeLabel:   strconv.Itoa(t.code),
+	}
+
+	i.counter.With(l).Inc()
+	elapsed := i.now().Sub(t.start)
+	i.duration.With(l).Observe(
+		float64(elapsed / time.Millisecond),
+	)
+
+	i.requestSize.With(l).Observe(
+		float64(t.requestSize),
+	)
+
+	if i.errorCount != nil && t.err != nil {
+		i.errorCount.With(l).Inc()
+	}
+}
+
+// ServerBundle is a prebaked set of serverside HTTP metrics.  A ServerBundle
+// is used to create middleware in the form of a ServerInstrumenter.
+type ServerBundle struct {
+	bundle
+}
+
+// NewServerBundle constructs a bundle of HTTP  metrics using the given Factory.
+// This function should be called at most once for any given Factory, or a
+// prometheus.AlreadyRegisteredError will occur.
+//
+// If now is nil, time.Now is used.
+func NewServerBundle(f *touchstone.Factory, now func() time.Time) (sb ServerBundle, err error) {
+	sb.now = now
+	if sb.now == nil {
+		sb.now = time.Now
+	}
+
+	sb.counter, err = f.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: MetricServerRequestCount,
+			Help: metricHelp[MetricServerRequestCount],
+		}, ServerLabel, CodeLabel, MethodLabel,
+	)
+
+	if err == nil {
+		sb.duration, err = f.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: MetricServerRequestDuration,
+				Help: metricHelp[MetricServerRequestDuration],
```

Comment:
for the `HistogramOpts` is there a way for a consumer to supply the Buckets?

Reply:
Yes: https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#HistogramOpts

You can just fill out the struct more.  I'll add that to the example, just for clarity.

Reply:
I'm sorry this is a miscommunication on my part. I understand how to use custom `HistogramOpts` I am curious specifically for the Metric `MetricServerRequestDuration`. Since `DefBuckets` the max bucket is 10ms which is too small for most of xmidt use cases.

Reply:
I think this is good enough https://github.com/xmidt-org/webpa-common/blob/90256afe7c190705e30ed769814aa56475580f4b/server/metrics.go#L49

Reply:
```Buckets: []float64{0.0625, 0.125, .25, .5, 1, 5, 10, 20, 40, 80, 160},```

Reply:
Ok I'll add that.  I'll also file a feature issue to allow a client to supply *Opts structs for metrics in a bundle, which will allow custom buckets.  That's a feature I already wanted to add anyway.

Reply:
Note those buckets are in seconds so it should probably be.
```golang
Buckets: []float64{62.5, 125, 250, 500, 1000, 5000, 10000, 20000, 40000, 80000, 160000},
```

Reply:
Would we prefer duration in seconds by default?  I made it milliseconds in `themis` because many of the HTTP requests take less than 1 second, making the distribution wonky.  I could also make the duration a configurable thing too.